base/reduce.jl

```diff
@@ -4,6 +4,14 @@
 
 ###### Generic (map)reduce functions ######
 
+"""
+    Base.NoInit()
+
+Singleton used for the purpose of indicating to the [`reduce`](@ref) family of functions
+that no initial value has been specified for the reduction.
+"""
+struct NoInit; end
```

Comment:
I must say I'm not a fan of this. This feels like a slippery slope to singletons for every possible situation. Instead I think it makes sense to use `nothing` here, which has become pretty conventional. If someone wants to _actually_ have `nothing` as their initial value, they should use `Some(nothing)`.

Reply:
Yeah - I'm definitely in two minds about it!

Reply:
Though, were you suggesting we unwrap `Something` here?

Reply:
Yeah. So `init=nothing` would mean "no initial value provided," whereas `init=Some(nothing)` would mean "the initial value is provided and is `nothing`."

Reply:
OK - bear with me, I'm trying to get my head around this pattern. If `init = Some(1)` do we use `1` or `Some(1)` for the inital value? How do I provide `Some(nothing)` as the initial element if I need to do that? I'll also need to find the best place to handle the unwrapping elegantly. Do we have a convenience function for unwrapping zero or one `Some`s?

Reply:
I guess that would be `Some(Some(nothing))` if you need actually need `Some(nothing)` as your initial element. As for unwrapping, take a look at the (admittedly oddly named) `something` function, which will return the first argument that is not `nothing`, and will unwrap `Some(nothing)`.

Reply:
I think this is a good idea and I'm in favour of more singletons like this. What's the harm if a user never has to create it and never sees it returned? Whereas the `Some` alternatives require these unwrapping gymnastics.

Reply:
I don't think they're unwrapping gymnastics; it's like escaping backslashes or escaping `$`; we use `nothing` to denote that the user has not passed anything in, so if the user for some reason wants to use `nothing` as a salient value itself, you need someway to "un-do the nothingness of `nothing`", which is to wrap it in `Something()`.

It is like asking the question "What if I have an array structures and I'm doing a reduction on them, and I want my initial value to literally be a `NoInit` structure?"  You need some way to literally represent a `NoInit` structure that should be used as the initial value of the reduction instead of interpreting that as the user not having passed anything in as the initializer.

The gymnastics only get introduced when you want to pass in escaped escaped values; e.g. if, for some reason, you wanted to start out with a `Something(nothing)` value, so you'd have to pass in `Something(Something(nothing))`.

I think the value here is in providing a coherent interface for all of `Base` and `stdlib`.  We now have an unambiguous way of telling a function "this is the default value" so we should use it instead of creating ad-hoc replacements for `nothing`.

Reply:
The counteragument is this can very easily end up like bash, where you really need to add `""` everywhere to be safe & generic (for us the anaolgy would be you need to add `Some(...)` to everything *just in case it has some special meaning you didn't predict*).

Reply:
You really only ever need to worry about wrapping with `Some` if you're dealing explicitly with `nothing`s, which should be an incredibly uncommon case.

Reply:
While I'm not going to express any preference towards using kwargs or not for this, if we do this, let's do without any special magic constants. Here's pseudo code for it (assuming I approximately get the definition of `mapfoldl` right):
```julia
function mapfoldl(f, op, itr; init...)
    if length(init) == 1
        v = init.init
        init = NamedTuple()
    end
    isempty(init) || throw(ArgumentError("incorrect kwarg")) # MethodError?
    for x in itr
        x = f(x)
        if @isdefined v
            v = op(v, x)
        else
            v = x
        end
    end
    return v
end
```

Reply:
Thanks Jameson, I like it!

Reply:
Yes, it will be a slightly messier implementation but it’s simpler for end users, which is what gains my vote here. 

It does make me wonder if the “pattern” could be supported in a more straightforward way by the language.